src/batch/src/executor2/join/nested_loop_join.rs

```diff
@@ -89,83 +91,67 @@ enum NestedLoopJoinState {
     Done,
 }
 
-#[async_trait::async_trait]
-impl Executor for NestedLoopJoinExecutor {
-    async fn open(&mut self) -> Result<()> {
-        match self.state {
-            NestedLoopJoinState::Build => {
-                // TODO: Do not all read to memory if OOM.
-                self.build_table.load_data().await?;
-                self.state = NestedLoopJoinState::FirstProbe;
-            }
-            _ => unreachable!(),
-        }
-        Ok(())
+impl Executor2 for NestedLoopJoinExecutor2 {
+    fn schema(&self) -> &Schema {
+        &self.schema
     }
 
-    async fn next(&mut self) -> Result<Option<DataChunk>> {
-        if let Some(last_chunk) = self.last_chunk.take() {
-            let (left_data_chunk, return_chunk) = self.chunk_builder.append_chunk(last_chunk)?;
-            self.last_chunk = left_data_chunk;
-            if let Some(return_chunk_inner) = return_chunk {
-                return Ok(Some(return_chunk_inner));
-            }
-        }
-        // Make sure all buffered chunk are flushed before generate more chunks.
-        // It's impossible that: left chunk not null and return chunk is null.
-        assert!(self.last_chunk.is_none());
+    fn identity(&self) -> &str {
+        &self.identity
+    }
 
-        loop {
-            match self.state {
-                NestedLoopJoinState::Build => unreachable!(),
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
 
+impl NestedLoopJoinExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        let mut state = NestedLoopJoinState::Build;
+        loop {
+            match state {
+                NestedLoopJoinState::Build => {
+                    self.build_table.load_data().await?;
+                    state = NestedLoopJoinState::FirstProbe;
+                }
                 NestedLoopJoinState::FirstProbe => {
-                    let ret = self.probe(true).await?;
-                    self.state = NestedLoopJoinState::Probe;
+                    let ret = self.probe(true, &mut state).await?;
+                    if state == NestedLoopJoinState::FirstProbe {
+                        state = NestedLoopJoinState::Probe;
+                    }
                     if let Some(data_chunk) = ret {
-                        return Ok(Some(data_chunk));
+                        yield data_chunk;
                     }
                 }
                 NestedLoopJoinState::Probe => {
-                    let ret = self.probe(false).await?;
+                    let ret = self.probe(false, &mut state).await?;
```

Comment:
What's the reason not use state as self member fields like before?

Reply:
I refer to this [pr](https://github.com/singularity-data/risingwave/pull/2119#discussion_r857426526) and consider that there is no need of a state machine to implement stream, so I put state as a local variable